eth_account/hdaccount/deterministic.py

```diff
@@ -0,0 +1,211 @@
+        let I = HMAC-SHA512(Key = c_par, Data = 0x00 || ser_256(k_par) || ser_32(i)).
+        (Note: The 0x00 pads the private key to make it 33 bytes long.)
+      - If not (normal child):
+        let I = HMAC-SHA512(Key = c_par, Data = ser_P(point(k_par)) || ser_32(i)).
+    - Split I into two 32-byte sequences, I_L and I_R.
+    - The returned child key k_i is parse_256(I_L) + k_par (mod n).
+    - The returned chain code c_i is I_R.
+    - In case parse_256(I_L) ≥ n or k_i = 0, the resulting key is invalid,
+      and one should proceed with the next value for i.
+      (Note: this has probability lower than 1 in 2**127.)
+    """
+    assert len(parent_chain_code) == 32
+    if isinstance(node, HardNode):
+        # NOTE Empty byte is added to align to SoftNode case
+        assert len(parent_key) == 32  # Should be guaranteed here in return statment
+        child = hmac_sha512(parent_chain_code, b"\x00" + parent_key + node.serialize())
+
+    else:
+        assert len(ec_point(parent_key)) == 33  # Should be guaranteed by Account class
+        child = hmac_sha512(parent_chain_code, ec_point(parent_key) + node.serialize())
+
+    assert len(child) == 64
+
+    if to_int(child[:32]) >= SECP256K1_N:
+        # Invalid key, compute using next node (< 2**-127 probability)
+        return derive_child_key(parent_key, parent_chain_code, node + 1)
+
+    child_key = (to_int(child[:32]) + to_int(parent_key)) % SECP256K1_N
+    if child_key == 0:
+        # Invalid key, compute using next node (< 2**-127 probability)
+        return derive_child_key(parent_key, parent_chain_code, node + 1)
+
+    child_key = child_key.to_bytes(32, byteorder="big")
+    child_chain_code = child[32:]
+    return child_key, child_chain_code
+
+
+class HDPath:
+    def __init__(self, path: str):
+        """
+        Constructor for this class. Initializes an hd account generator using the
+        given path string (from BIP-0032). The path is decoded into nodes of the
+        derivation key tree, which define a pathway from a given master seed to
+        the child key that is used for a given purpose. Please also reference BIP-
+        0043 (which definites the first level as the "purpose" field of an HD path)
+        and BIP-0044 (which defines a commonly-used, 5-level scheme for BIP32 paths)
+        for examples of how this object may be used. Please note however that this
+        object makes no such assumptions of the use of BIP43 or BIP44, or later BIPs.
+        :param path             : BIP32-compatible derivation path
+        :type path              : str as "m/idx_0/.../idx_n" or "m/idx_0/.../idx_n"
+                                  where idx_* is either an integer value (soft node)
+                                  or an integer value followed by either the "'" char
+                                  or the "H" char (hardened node)
+        """
+        nodes = path.split('/')
+        if nodes[0] not in ('m', 'M'):
```

Comment:
If we want to be really careful we should pre-validate that `path` is not the empty string, otherwise this will raise an `IndexError` since `nodes` will be the empty list.

Reply:
Also, can you move `('m', 'M')` to be a module level constant and make it a `set` since it is just used for inclusion checks.

Reply:
Will do